File: 3510. Minimum Pair Removal to Sort Array II/Solution.cpp
#include <numeric>
#include <queue>
#include <vector>
using namespace std;

class Solution {
public:
    int minimumPairRemoval(vector<int>& nums)
    {
        int n = nums.size(), dec = 0;
        priority_queue<pair<long long, int>, vector<pair<long long, int>>, greater<>> pq;
        for (int i = 0; i < n - 1; ++i) {
            if (nums[i] > nums[i + 1])
                ++dec;
            pq.emplace(nums[i] + nums[i + 1], i);
        }

        vector<int> prev(n), next(n);
        iota(prev.begin(), prev.end(), -1);
        iota(next.begin(), next.end(), 1);

        int result = 0;
        vector<long long> A(nums.begin(), nums.end());
        while (dec) {
            auto [s, i] = pq.top();
            pq.pop();
            if (next[i] >= n || 0LL + A[i] + A[next[i]] != s)
                continue;
            ++result;
            int prev_i = prev[i];
            if (prev_i > -1) {
                dec -= A[prev_i] > A[i];
                dec += A[prev_i] > s;
                pq.emplace(A[prev_i] + s, prev_i);
            }
            int next_i = next[i];
            dec -= A[i] > A[next_i];
            int next_next_i = next[next_i];
            if (next_next_i < n) {
                dec -= A[next_i] > A[next_next_i];
                dec += s > A[next_next_i];
                prev[next_next_i] = i;
                pq.emplace(s + A[next_next_i], i);
            }
            A[i] = s;
            next[i] = next_next_i;
            next[next_i] = n;
        }
        return result;
    }
};

File: 3510. Minimum Pair Removal to Sort Array II/Solution.py
import heapq
from itertools import pairwise


class Solution:
    def minimumPairRemoval(self, nums: list[int]) -> int:
        pq = []
        dec = 0
        for i, (x, y) in enumerate(pairwise(nums)):
            if x > y:
                dec += 1
            pq.append((x + y, i))
        heapq.heapify(pq)

        n = len(nums)
        prev_list = list(range(-1, n - 1))
        next_list = list(range(1, n + 1))

        result = 0
        while dec:
            s, i = heapq.heappop(pq)
            next_i = next_list[i]
            if next_i >= n or nums[i] + nums[next_i] != s:
                continue
            result += 1

            prev_i = prev_list[i]
            if prev_i > -1:
                dec -= nums[prev_i] > nums[i]
                dec += nums[prev_i] > s
                heapq.heappush(pq, (nums[prev_i] + s, prev_i))

            dec -= nums[i] > nums[next_i]

            next_next_i = next_list[next_i]
            if next_next_i < n:
                dec -= nums[next_i] > nums[next_next_i]
                dec += s > nums[next_next_i]
                prev_list[next_next_i] = i
                heapq.heappush(pq, (s + nums[next_next_i], i))

            nums[i] = s
            next_list[i] = next_next_i
            next_list[next_i] = n
        return result
